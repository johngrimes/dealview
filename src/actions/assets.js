// @flow

import { putObject, deleteObject, getAllObjects } from 'db/db'
import { invalidateBalanceSheet } from 'actions/balanceSheet'
import type { Asset, AssetWithId, AssetMap } from 'types/assets/asset'
import type { Thunk } from 'types/commonTypes'

const objectStore = 'Asset'

type PutAssetRequestAction = {
  type: 'PUT_ASSET_REQUEST',
  asset: Asset,
}
type PutAssetSuccessAction = {
  type: 'PUT_ASSET_SUCCESS',
  asset: AssetWithId
}
type PutAssetFailureAction = {
  type: 'PUT_ASSET_FAILURE',
  error: string|null
}

type DeleteAssetRequestAction = {
  type: 'DELETE_ASSET_REQUEST',
  id: string,
}
type DeleteAssetSuccessAction = {
  type: 'DELETE_ASSET_SUCCESS',
  id: string
}
type DeleteAssetFailureAction = {
  type: 'DELETE_ASSET_FAILURE',
  error: string|null
}

type LoadAssetsRequestAction = {
  type: 'LOAD_ASSETS_REQUEST'
}
type LoadAssetsSuccessAction = {
  type: 'LOAD_ASSETS_SUCCESS',
  assets: AssetMap
}
type LoadAssetsFailureAction = {
  type: 'LOAD_ASSETS_FAILURE',
  error: string|null
}

export type AssetAction = PutAssetRequestAction
                        | PutAssetSuccessAction
                        | PutAssetFailureAction
                        | DeleteAssetRequestAction
                        | DeleteAssetRequestAction
                        | DeleteAssetFailureAction
                        | LoadAssetsRequestAction
                        | LoadAssetsSuccessAction
                        | LoadAssetsFailureAction

export const putAssetRequest = (asset: Asset): PutAssetRequestAction => {
  return {
    type: 'PUT_ASSET_REQUEST',
    asset,
  }
}

export const putAssetSuccess = (asset: AssetWithId): PutAssetSuccessAction => {
  return {
    type: 'PUT_ASSET_SUCCESS',
    asset,
  }
}

export const putAssetFailure = (error: string|null): PutAssetFailureAction => {
  return {
    type: 'PUT_ASSET_FAILURE',
    error,
  }
}

export const deleteAssetRequest = (id: string): DeleteAssetRequestAction => {
  return {
    type: 'DELETE_ASSET_REQUEST',
    id,
  }
}

export const deleteAssetSuccess = (id: string): DeleteAssetSuccessAction => {
  return {
    type: 'DELETE_ASSET_SUCCESS',
    id,
  }
}

export const deleteAssetFailure = (error: string|null): DeleteAssetFailureAction => {
  return {
    type: 'DELETE_ASSET_FAILURE',
    error,
  }
}

export const loadAssetsRequest = (): LoadAssetsRequestAction => {
  return { type: 'LOAD_ASSETS_REQUEST' }
}

export const loadAssetsSuccess = (assets: AssetMap): LoadAssetsSuccessAction => {
  return {
    type: 'LOAD_ASSETS_SUCCESS',
    assets,
  }
}

export const loadAssetsFailure = (error: string|null): LoadAssetsFailureAction => {
  return {
    type: 'LOAD_ASSETS_FAILURE',
    error,
  }
}

export const putAsset = (asset: Asset): Thunk => {
  return dispatch => {
    dispatch(putAssetRequest(asset))
    return new Promise(async (resolve, reject) => {
      try {
        const saved = await putObject(objectStore, asset)
        dispatch(invalidateBalanceSheet())
        dispatch(putAssetSuccess(saved))
        resolve(saved)
      } catch (error) {
        dispatch(putAssetFailure(error))
        reject(error)
      }
    })
  }
}

export const deleteAsset = (id: string): Thunk => {
  return dispatch => {
    dispatch(deleteAssetRequest(id))
    return new Promise(async (resolve, reject) => {
      try {
        const key = await deleteObject(objectStore, id)
        dispatch(invalidateBalanceSheet())
        dispatch(deleteAssetSuccess(key))
        resolve(key)
      } catch (error) {
        dispatch(deleteAssetFailure(error))
        reject(error)
      }
    })
  }
}

export const loadAssets = (): Thunk => {
  return dispatch => {
    dispatch(loadAssetsRequest())
    return new Promise(async (resolve, reject) => {
      try {
        const assets = await getAllObjects(objectStore)
        dispatch(loadAssetsSuccess(assets))
        resolve(assets)
      } catch (error) {
        dispatch(loadAssetsFailure(error))
        reject(error)
      }
    })
  }
}
