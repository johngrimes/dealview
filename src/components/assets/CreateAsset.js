// @flow

import React from 'react'
import type { Children } from 'react'

const CreateAsset = (props: { children: Children }) => {
  const { children } = props

  return (
    <div className='create-asset'>
      <h1>Create Asset</h1>
      {children}
    </div>
  )
}

export default CreateAsset
