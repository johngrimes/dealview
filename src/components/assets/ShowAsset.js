// @flow

import React from 'react'
import type { Children } from 'react'

const ShowAsset = (props: { children: Children }) => {
  const { children } = props

  return (
    <div className="show-asset">
      <h1>Show Asset</h1>
      {children}
    </div>
  )
}

export default ShowAsset
