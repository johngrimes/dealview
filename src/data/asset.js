import moment from 'moment'

export const getValuationAtDate = (asset, date) => {
  const now = moment().startOf('day')
  const startDate = asset.purchaseDate || now
  if (
    date.isBefore(startDate) ||
    (asset.saleDate && date.isAfter(asset.saleDate))
  ) {
    return 0
  }
  const daysSinceStart = Math.floor(
    moment.duration(date.diff(startDate)).asDays()
  )
  const startValue = asset.purchaseAmount
    ? parseInt(asset.purchaseAmount, 10)
    : parseInt(asset.currentValue, 10)
  const forecastGrowth = parseInt(asset.forecastGrowth, 10) / 10000 / 365
  return Math.round(startValue * Math.pow(1 + forecastGrowth, daysSinceStart))
}
