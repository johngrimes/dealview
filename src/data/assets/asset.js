// @flow

export type Asset = {
  type: 'RealEstate',
  instanceId: string,
  name: string,
  lastValuation?: number
}

export type AssetWithId = Asset & { id: string }

export type AssetMap = { [id: string]: AssetWithId }
