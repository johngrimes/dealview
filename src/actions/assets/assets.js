// @flow

import { putObject, deleteObject, getAllObjects } from 'db/db'
import { invalidateBalanceSheet } from 'actions/balanceSheet'
import { createObjectActions } from 'actions/objects'
import type { Asset } from 'types/assets/asset'
import type { Thunk } from 'types/commonTypes'

const objectStore = 'Asset'

const actions = createObjectActions('ASSET', 'ASSETS',
                                    'Asset', 'Assets')

const putAsset = (asset: Asset): Thunk => {
  return dispatch => {
    dispatch(actions.putAssetRequest(asset))
    return new Promise(async (resolve, reject) => {
      try {
        const saved = await putObject(objectStore, asset)
        dispatch(invalidateBalanceSheet())
        dispatch(actions.putAssetSuccess(saved))
        resolve(saved)
      } catch (error) {
        dispatch(actions.putAssetFailure(error))
        reject(error)
      }
    })
  }
}

const deleteAsset = (id: string): Thunk => {
  return dispatch => {
    dispatch(actions.deleteAssetRequest(id))
    return new Promise(async (resolve, reject) => {
      try {
        const key = await deleteObject(objectStore, id)
        dispatch(invalidateBalanceSheet())
        dispatch(actions.deleteAssetSuccess(key))
        resolve(key)
      } catch (error) {
        dispatch(actions.deleteAssetFailure(error))
        reject(error)
      }
    })
  }
}

const loadAssets = (): Thunk => {
  return dispatch => {
    dispatch(actions.loadAssetsRequest())
    return new Promise(async (resolve, reject) => {
      try {
        const assets = await getAllObjects(objectStore)
        dispatch(actions.loadAssetsSuccess(assets))
        resolve(assets)
      } catch (error) {
        dispatch(actions.loadAssetsFailure(error))
        reject(error)
      }
    })
  }
}

export default {
  ...actions,
  putAsset,
  deleteAsset,
  loadAssets,
}
