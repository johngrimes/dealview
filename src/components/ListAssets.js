import React from 'react'
import { connect } from 'react-redux'
import _ from 'lodash'
import { Link } from 'react-router-dom'
import moment from 'moment'

import Breadcrumbs from './Breadcrumbs.js'
import RealEstateActions from '../actions/assets.js'
import { getValuationAtDate } from '../data/asset.js'
import { formatDollars } from '../data/commonTypes.js'

export class ListAssets extends React.Component {
  constructor(props) {
    super(props)
    if (this.props.assets.status === 'uninitialised') {
      this.props.dispatch(RealEstateActions.loadAssets())
    }
  }

  breadcrumbs() {
    return [
      { display: 'Portfolio', path: '/portfolio' },
      { display: 'Assets', path: '/portfolio/assets' },
    ]
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.assets.status === 'uninitialised') {
      this.props.dispatch(RealEstateActions.loadAssets())
    }
  }

  render() {
    const assets = _.map(this.props.assets.objects, (v, k) => {
      const lastValuation = getValuationAtDate(v, moment())
      const lastValuationTag =
        lastValuation === 0
          ? null
          : <div className='asset-last-valuation'>
              {formatDollars(lastValuation)}
            </div>
      return (
        <li key={k} className='asset'>
          <Link
            className='asset-name'
            to={`/portfolio/assets/real-estate/${v.id}`}
          >
            {v.name}
          </Link>
          {lastValuationTag}
        </li>
      )
    })
    return (
      <div className='list-assets'>
        <Breadcrumbs breadcrumbs={this.breadcrumbs()} />
        <p>
          <Link to='/portfolio/assets/new'>New Asset</Link>
        </p>
        <ol className='assets'>
          {assets}
        </ol>
      </div>
    )
  }
}

const mapStateToProps = state => ({ assets: state.assets })

export default connect(mapStateToProps)(ListAssets)
