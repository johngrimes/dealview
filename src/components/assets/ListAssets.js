// @flow

import React from 'react'
import { connect } from 'react-redux'
import _ from 'lodash'
import { Link } from 'react-router-dom'
import moment from 'moment'
import type { Dispatch } from 'redux'

import Breadcrumbs from 'components/Breadcrumbs/Breadcrumbs'
import { loadAssets } from 'actions/assets'
import { getValuationAtDate } from 'types/assets/asset'
import { DateFormat, formatDollars } from 'types/commonTypes'
import type { GlobalState } from 'store'
import type { AssetState } from 'reducers/assets'
import type { BreadcrumbTrail } from 'components/Breadcrumbs/Breadcrumbs'

type Props = {
  assets: AssetState,
  breadcrumbs: BreadcrumbTrail,
  dispatch: Dispatch
}

export class ListAssets extends React.Component {
  props: Props

  constructor(props: Props) {
    super(props)
    if (this.props.assets.status === 'uninitialised') {
      this.props.dispatch(loadAssets())
    }
  }

  breadcrumbs(): BreadcrumbTrail {
    return [
      { display: 'Portfolio', path: '/portfolio' },
      { display: 'Assets', path: '/portfolio/assets' },
    ]
  }

  componentWillReceiveProps(nextProps: Props) {
    if (nextProps.assets.status === 'uninitialised') {
      this.props.dispatch(loadAssets())
    }
  }

  render() {
    const assets = _.map(this.props.assets.objects, (v, k) => {
      const lastValuation = getValuationAtDate(v, moment().format(DateFormat))
      const lastValuationTag = lastValuation === 0
        ? null
        : <div className='asset-last-valuation'>{formatDollars(lastValuation)}</div>
      return <li key={k} className='asset'>
        <Link className='asset-name' to={`/portfolio/assets/real-estate/${v.id}`}>{v.name}</Link>
        {lastValuationTag}
      </li>
    })
    return (
      <div className='list-assets'>
        <Breadcrumbs breadcrumbs={this.breadcrumbs()} />
        <p><Link to='/portfolio/assets/new'>New Asset</Link></p>
        <ol className='assets'>
          {assets}
        </ol>
      </div>
    )
  }
}

const mapStateToProps = (state: GlobalState) => {
  return {
    assets: state.assets,
  }
}

export default connect(mapStateToProps)(ListAssets)
