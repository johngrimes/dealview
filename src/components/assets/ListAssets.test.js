/* global expect */

import { ListAssets } from './ListAssets.js'

import React from 'react'
import { shallow } from 'enzyme'

import { validAssetWithId1, validAssetWithId2 } from '../../test/fixtures/asset.js'
import { loadAssets } from '../../actions/assets.js'

describe('ListAssets', () => {
  it('should render some Assets', () => {
    const props = {
      assets: {
        status: 'loaded',
        objects: {
          [validAssetWithId1.id]: validAssetWithId1,
          [validAssetWithId2.id]: validAssetWithId2
        }
      }
    }
    const wrapper = shallow(<ListAssets {...props} />)
    expect(wrapper.find('.asset').length).toBe(Object.keys(props.assets).length)
    expect(wrapper).toMatchSnapshot()
  })

  it('should load Assets if uninitialised', () => {
    const dispatch = jest.fn()
    const assetActions = require('../../actions/assets.js')
    assetActions.loadAssets = jest.fn()
    const props = {
      assets: {
        status: 'uninitialised',
        objects: {}
      },
      dispatch
    }
    const wrapper = shallow(<ListAssets {...props} />)
    expect(dispatch).toHaveBeenCalled()
    expect(assetActions.loadAssets).toHaveBeenCalled()
  })
})
