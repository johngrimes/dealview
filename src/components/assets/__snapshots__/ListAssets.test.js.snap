// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListAssets should render some Assets 1`] = `
<div
  className="list-assets"
>
  <Breadcrumbs
    breadcrumbs={
      Array [
        Object {
          "display": "Portfolio",
          "path": "/portfolio",
        },
        Object {
          "display": "Assets",
          "path": "/portfolio/assets",
        },
      ]
    }
  />
  <p>
    <Link
      replace={false}
      to="/portfolio/assets/new"
    >
      New Asset
    </Link>
  </p>
  <ol
    className="assets"
  >
    <li
      className="asset"
    >
      <Link
        className="asset-name"
        replace={false}
        to="/portfolio/assets/real-estate/98"
      >
        5 Sunrise St
      </Link>
      <div
        className="asset-last-valuation"
      >
        705000
      </div>
    </li>
    <li
      className="asset"
    >
      <Link
        className="asset-name"
        replace={false}
        to="/portfolio/assets/real-estate/99"
      >
        38/8 Briggs Road
      </Link>
      <div
        className="asset-last-valuation"
      >
        260000
      </div>
    </li>
  </ol>
</div>
`;
